corehq/form_processor/backends/sql/dbaccessors.py

```diff
@@ -507,13 +511,24 @@ def save_new_form(form):
         for operation in operations:
             operation.form = form
 
-        with get_cursor(XFormInstanceSQL) as cursor:
-            cursor.execute(
-                'SELECT form_pk FROM save_new_form_and_related_models(%s, %s, %s, %s)',
-                [form.form_id, form, unsaved_attachments, operations]
-            )
-            result = fetchone_as_namedtuple(cursor)
-            form.id = result.form_pk
+        try:
+            with transaction.atomic(using=db_name):
+                form.save(using=db_name)
+                for attachment in unsaved_attachments:
+                    if attachment.is_saved():
+                        raise XFormSaveError(
+                            'XFormAttachmentSQL {} has already been saved'.format(attachment.id)
+                        )
+                    attachment.save(using=db_name)
+
+                for operation in operations:
+                    if operation.is_saved():
+                        raise XFormSaveError(
+                            'XFormOperationSQL {} has already been saved'.format(operation.id)
+                        )
+                    operation.save(using=db_name)
```

Comment:
I know this is in an atomic transaction but seems like we could check the attachments and operations before and only start the transaction of all is good.

I know we have the same pattern in the case save but just noticing it now.